Source/DafnyCore/Dafny.atg

```diff
@@ -4497,20 +4507,30 @@ QuantifierExpression<out Expression q, bool allowLemma, bool allowLambda>
      List<BoundVar/*!*/> bvars;
      Attributes attrs;
      Expression range;
-     Expression/*!*/ body;
+     q = dummyExpr;
   .)
   ( Forall                                     (. x = t;  univ = true; .)
   | Exists                                     (. x = t; .)
   )
-  QuantifierDomain<out bvars, out attrs, out range, allowLemma, allowLambda, true>
-  QSep
-  Expression<out body, allowLemma, allowLambda>
-  (. if (univ) {
-       q = new ForallExpr(x, new RangeToken(x, t), bvars, range, body, attrs);
-     } else {
-       q = new ExistsExpr(x, new RangeToken(x, t), bvars, range, body, attrs);
-     }
-  .)
+  QuantifierDomain<out bvars, out attrs, out range, true, true, true>
+  ( IF(univ && (la.kind == _ensures || la.kind == _lbrace))
+    // It's a forall statement. Parse it as part of a StmtExpr.
```

Comment:
That works. It means we could theoretically do the same for if statements (last week I would have love to have that) and I think that could be useful too and less surprising in the future, although out of scope for this PR.

Reply:
I agree it's out of scope for this PR, but I do find it an interesting idea. I suppose such an `if` statement would still just be for the `StmtExpr`, rather than being an alternative syntax for `if-then-else` expressions? This would then allow you to write some proofs inside the `if` statement and follow that by the expression you want to return. (I have never felt the need for this myself, but perhaps only because I knew the feature wasn't available.)

Some new users confuse `if-then-else` expressions and `if` statements (which, curiously, seems not to be a confusion in languages that use the more cryptic `? :` for the former). On one hand, also allowing `if` statements in expressions might add to that confusion. But on the other hand, maybe this would give us a place in the parser to give a better error message if the program gets this wrong.

While we're thinking about such things, one could also imagine other statements, in particular `while` loops, in expressions. Again, such a `while` statement would just be for the `StmtExpr`, to allow an inline proof that uses a loop.